pkg/registry/core/service/storage/rest.go

```diff
@@ -427,7 +428,8 @@ func (rs *REST) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 		releaseNodePorts(oldService, nodePortOp)
 	}
 	// Update service from any type to NodePort or LoadBalancer, should update NodePort.
-	if service.Spec.Type == api.ServiceTypeNodePort || service.Spec.Type == api.ServiceTypeLoadBalancer {
+	if service.Spec.Type == api.ServiceTypeNodePort ||
+		(service.Spec.Type == api.ServiceTypeLoadBalancer && (service.Spec.AllocateLoadBalancerNodePorts == nil || *service.Spec.AllocateLoadBalancerNodePorts)) {
```

Comment:
Rather than test for nil, you can add a small block to pkg/apis/core/v1/defaults.go which checks for nil and sets it "true" if the gate allows it to be set.

Then here you can do something like:

```
allocateNodePorts := shouldAllocateNodePorts(service)
```

and

```
func shouldAllocateNodePorts(service) bool {
    if gate disabled {
        return true
    }
    return service.spec.AllocateLoadBalancerNodePorts
}
```

Reply:
Update and Create, of course